### ISSUE TITLE
error 'terminate called after throwing an instance of 'std::length_error'' after unclean shutdown

### ISSUE THREAD

**Issue by whysthatso** [sigp/lighthouse#3458]
## Description

when starting with the following config:

```
lighthouse --network mainnet  --datadir /data/lighthouse bn --debug-level=debug
```
i'm receiving this output with a subsequent abort:

```
INFO Logging to file                         path: "/data/lighthouse/beacon/logs/beacon.log"
INFO Lighthouse started                      version: Lighthouse/v2.3.1-564d7da
INFO Configured for network                  name: mainnet
INFO Data directory initialised              datadir: /data/lighthouse
INFO Deposit contract                        address: 0x00000000219ab540356cbb839cbe05303d7705fa, deploy_block: 11184524
DEBG Ignoring slots-per-restore-point config in favour of on-disk value, on_disk: 2048, config: 8192, service: freezer_db
INFO Hot-Cold DB initialized                 split_state: 0x2b60c9783c6d6bb91e7abd69e13abea1cfec1cead0f55c936dcc6f92690a7f1b, split_slot: 4339808, service: freezer_db
DEBG Attempting schema migration             to_version: 9, from_version: 9, service: freezer_db
INFO Starting beacon chain                   method: resume, service: beacon

terminate called after throwing an instance of 'std::length_error'
  what():  basic_string::_M_create
Aborted
```
i'm curious if this could be caused by data corruption due to an unclean shutdown?

## Version
Lighthouse/v2.3.1-564d7da
this is running on archlinux, source is AUR: https://aur.archlinux.org/packages/lighthouse-ethereum

**Comment by whysthatso** [comment]
removing the datadir fixed the problem, so i assume it's a data corruption issue. what would be the best practice to possibly fix data corruption in the future? any tools for that?

**Comment by michaelsproul** [comment]
AFAIK no tools exist to repair corrupt LevelDB databases. It's an unfortunate reality that a re-sync is the best option if this happens.

What filesystem were you running Lighthouse on? Did you have a power outage that could have corrupted the DB?

**Comment by whysthatso** [comment]
@michaelsproul i've asked a related, yet more general question in the recent reddit AMA [here](https://www.reddit.com/r/ethereum/comments/wpqmo1/comment/iki5qb4/)

server locked up due to power issues, yes.
i've played around with deleting the last bunch of block files, and putting the pointer to a known complete block, deleting the manifest. that caused the db driver (?) to recheck all blocks (could be followed in the LOG file inside the chainstate db directory, but still led to the same issue. the LOG file showed some other error, unfortunately, i have not recorded that one. something along the lines of 'this value is too short' and then something in hex.

i know this is a bit hazy of a report, but i guess i am also not expecting too much of a resolution. it's more for posterity.

the TLDR for the reddit answer was that corruption should not occur unless there is an underlying hardware issue. i'm running this on a arch linux vm on proxmox, fs is ext4, the underlaying ssd volumes are on a hardware-based raid5 system. the controller does not show any hardware issues, so it's difficult to imagine that being the problem, but i can also not rule it out.
